File: apps/api/src/api/models/analytics.py
"""Request/response models for analytics API — Phase C.2 dashboard."""

from __future__ import annotations

from pydantic import BaseModel


class SummaryResponse(BaseModel):
    """Response for GET /analytics/summary."""

    total: int
    resolved: int
    unresolved: int
    avg_resolution_days: float | None = None


class CategoryStatsResponse(BaseModel):
    """Single category entry for GET /analytics/by-category."""

    category: str
    count: int
    percentage: float
    avg_severity: float
    avg_resolution_days: float | None = None


class TeamStatsResponse(BaseModel):
    """Single team entry for GET /analytics/by-team."""

    team: str
    count: int
    top_categories: list[str]
    avg_resolution_days: float | None = None


class TimelinePointResponse(BaseModel):
    """Single weekly point for GET /analytics/timeline."""

    week: str  # ISO date string
    count: int
    by_category: dict[str, int]


class SeverityTrendResponse(BaseModel):
    """Single weekly point for GET /analytics/severity-trend."""

    week: str  # ISO date string YYYY-MM-DD
    error_count: int
    warning_count: int


class RuleEffectivenessResponse(BaseModel):
    """Single rule entry for GET /analytics/top-rules."""

    rule_id: str
    incident_count: int
    avg_severity: float  # 0.5 (warning) or 1.0 (error)

File: apps/web/src/lib/types/analytics.ts
/** TypeScript types for incident analytics dashboard — Spec-014. */

export type Period = 'week' | 'month' | 'quarter' | 'custom';
export type StatusFilter = 'resolved' | 'unresolved' | 'all';
export type RootCauseCategory =
	| 'code_pattern'
	| 'infrastructure'
	| 'process_breakdown'
	| 'third_party'
	| 'unknown';

export interface AnalyticsFilter {
	period: Period;
	teams?: string[]; // Multi-select: ?team=a&team=b; empty = no filter
	category?: RootCauseCategory | null;
	status: StatusFilter;
	start_date?: string | null; // YYYY-MM-DD (required if period=custom)
	end_date?: string | null; // YYYY-MM-DD (required if period=custom)
}

export interface AnalyticsSummary {
	total: number;
	resolved: number;
	unresolved: number;
	avg_resolution_days: number | null;
}

export interface CategoryStats {
	category: RootCauseCategory;
	count: number;
	percentage: number; // 0–100
	avg_severity: number; // 0.5 (warning) or 1.0 (error)
	avg_resolution_days: number | null;
}

export interface TeamStats {
	team: string;
	count: number;
	top_categories: RootCauseCategory[];
	avg_resolution_days: number | null;
}

export interface TimelinePoint {
	week: string; // ISO date string YYYY-MM-DD
	count: number;
	by_category: Record<RootCauseCategory, number>; // Always all 5 keys
}

export interface SeverityTrendPoint {
	week: string; // ISO date string YYYY-MM-DD
	error_count: number;
	warning_count: number;
}

export interface RuleEffectivenessStats {
	rule_id: string;
	incident_count: number;
	avg_severity: number; // 0.5 (warning) or 1.0 (error)
}
